rclc/src/rclc/multi_threaded_executor.c

```diff
@@ -0,0 +1,530 @@
+          return rc;
+        }
+        break;
+
+
+      case RCLC_CLIENT:
+        // add client to wait_set and save index
+        rc = rcl_wait_set_add_client(
+          &executor->wait_set, executor->handles[i].client,
+          &executor->handles[i].index);
+        if (rc == RCL_RET_OK) {
+          RCUTILS_LOG_DEBUG_NAMED(
+            ROS_PACKAGE_NAME, "Client added to wait_set_client[%ld]",
+            executor->handles[i].index);
+        } else {
+          PRINT_RCLC_ERROR(rclc_executor_rebuild_wait_set, rcl_wait_set_add_client);
+          return rc;
+        }
+        break;
+
+      case RCLC_GUARD_CONDITION:
+        // add guard_condition to wait_set and save index
+        rc = rcl_wait_set_add_guard_condition(
+          &executor->wait_set, executor->handles[i].gc,
+          &executor->handles[i].index);
+        if (rc == RCL_RET_OK) {
+          RCUTILS_LOG_DEBUG_NAMED(
+            ROS_PACKAGE_NAME, "Guard_condition added to wait_set_client[%ld]",
+            executor->handles[i].index);
+          printf("added guard condition to index %ld\n", executor->handles[i].index);
+        } else {
+          PRINT_RCLC_ERROR(rclc_executor_rebuild_wait_set, rcl_wait_set_add_guard_condition);
+          return rc;
+        }
+        break;
+
+      default:
+        RCUTILS_LOG_DEBUG_NAMED(
+          ROS_PACKAGE_NAME, "Error: unknown handle type: %d",
+          executor->handles[i].type);
+        PRINT_RCLC_ERROR(rclc_executor_rebuild_wait_set, rcl_wait_set_add_unknown_handle);
+        return RCL_RET_ERROR;
+    }
+  }
+  return RCL_RET_OK;
+}
+void * rclc_executor_worker_thread(void * p)
+{
+  rclc_executor_worker_thread_param_t * param = (rclc_executor_worker_thread_param_t *)p;
+
+  // print priority
+  struct sched_param sp;
+  int result;
+  result = sched_getparam(0, &sp);
+  if (result < 0) {
+    printf("uros_rbs: sched_getparam failed: %d\n", result);
+    // todo (jast) is * void necessary?
+    return p;
+  }
+  printf("worker thread with prio %d\n", sp.sched_priority);
+
+  // endless worker_thread loop
+  while (1) {
+    // printf("worker-thread %ld.\n",param->handle->index); // not thread-safe
+    pthread_mutex_lock(&param->handle->new_msg_mutex);
+    while (!param->handle->new_msg_avail) {
+      // printf("worker thread: idling\n");
+      pthread_cond_wait(&param->handle->new_msg_cond, &param->handle->new_msg_mutex);
+    }
+    param->handle->new_msg_avail = false;
+    pthread_mutex_unlock(&param->handle->new_msg_mutex);
+
+    // execute subscription callback
+    // pthread_mutex_lock(param->micro_ros_mutex);
+    param->handle->subscription_callback(param->handle->data);
+    // pthread_mutex_unlock(param->micro_ros_mutex);
+
+    // change_worker thread state and signal guard condition
+    rclc_executor_change_worker_thread_state(param, RCLC_THREAD_READY);
+  }
+  // only for linters
+  return p;
+}
+
+rcl_ret_t
+rclc_executor_spin_multi_threaded(rclc_executor_t * e)
```

Comment:
I would prefer `executor` over `e` for readability. It is not used too often and will likely not cause additional line breaks. But that's a matter of taste.

Reply:
resolved in [e4f6226](https://github.com/ros2/rclc/pull/87/commits/e4f6226172236b975f9eea3262f8a4ba9e5a0d30)